Lua/Lua.sublime-syntax

```diff
@@ -757,6 +814,442 @@ contexts:
+    - include: pattern-unquoted-charclass
+
+  pattern-unquoted-set-ranges:
+    - meta_include_prototype: false
+    # These only work with direct characters. Lua doesn't support percent escapes in sets
+    - match: (?=[^%\]\-]-[^\]])
+      push:
+        - pattern-unquoted-set-range-end
+        - pattern-unquoted-set-range-begin
+
+  pattern-unquoted-set-range-begin:
+    - meta_include_prototype: false
+    - match: '-'
+      scope: punctuation.separator.sequence.lua
+      pop: 1
+
+  pattern-unquoted-set-range-end:
+    - meta_include_prototype: false
+    - meta_scope: constant.other.range.lua
+    - match: .
+      pop: 1
+
+  # this gets included in expressions that want to have a pattern string
+  pattern-string:
+    - match: \[(=*)\[
+      scope: punctuation.definition.string.begin.lua
+      set:
+        - pattern-multiline-string-body
+        - pattern-test-first-caret-multiline
+    - match: \'
+      scope: punctuation.definition.string.begin.lua
+      set:
+        - pattern-single-quoted-string-body
+        - pattern-test-first-caret
+    - match: \"
+      scope: punctuation.definition.string.begin.lua
+      set:
+        - pattern-double-quoted-string-body
+        - pattern-test-first-caret
+
+  pattern-test-first-caret:
+    - meta_include_prototype: false
+    - match: (\^)[+*?\-]?
+      captures:
+        1: keyword.control.anchor.lua
+      pop: 1
+
+    # eat these because they can't be a repetition at the start
+    - match: '[+*?\-]'
+      pop: 1
+    # it seems to require this match rather than immediately-pop
+    # in order to work with the multiline below
+    - match: (?=[^\^])
+      pop: 1
+
+  pattern-test-first-caret-multiline:
+    - meta_include_prototype: false
+    - match: $\n?
+      set: pattern-test-first-caret
+    - include: pattern-test-first-caret
+
+  pattern-single-quoted-string-body:
+    - meta_include_prototype: false
+    - meta_scope: meta.string.lua string.quoted.single.lua
+    - include: single-quoted-string-body
+    - match: \$(?=\')
+      scope: keyword.control.anchor.lua
+    - match: ''
+      push: pattern-quoted
+      with_prototype:
+        - match: (?=\$?\')
+          pop: 1
+
+  pattern-double-quoted-string-body:
+    - meta_include_prototype: false
+    - meta_scope: meta.string.lua string.quoted.double.lua
+    - include: double-quoted-string-body
+    - match: \$(?=\")
+      scope: keyword.control.anchor.lua
+    - match: ''
+      push: pattern-quoted
+      with_prototype:
+        - match: (?=\$?\")
+          pop: 1
+
+  pattern-multiline-string-body:
+    - meta_include_prototype: false
+    - meta_scope: meta.string.lua string.quoted.multiline.lua
+    - include: multiline-string-body
+    - match: ''
+      embed: pattern-unquoted
+      # we don't seem to be able to use the capture to get the right one
+      # this means that patterns that contain something that matches this
+      # will lose their context half way through.
+      escape: (?=\]=*\])
```

Comment:
Changes to simple multi-push, optionally consuming `\n` directly after opening punctuation.

```suggestion
  pattern-multiline-string-body:
    - meta_include_prototype: false
    - match: \n?
      push:
        - pattern-unquoted
        - pattern-test-first-caret
```